### PR TITLE
[Feature] support device name, and hide receiver when bluetooth active

### PR DESCRIPTION
**Pull request by ahammer**
## Purpose

* Hides "Receiver/Android" when Bluetooth is connected, and shows the Bluetooth device name

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

* Test the code

```
Connect a bluetooth device, verify "Receiver/Android" disapears, Verify correct device name is shown.
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/calling/CallingViewModel.kt [hunk 1]
```diff
@@ -148,6 +148,7 @@ internal class CallingViewModel(
 
         audioDeviceListViewModel.update(
             state.localParticipantState.audioState.device,
+            state.localParticipantState.audioState.bluetoothDeviceName,
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/calling/CallingViewModel.kt:151]: Should we consider to use a bluetooth dedicated class in the audioState?

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt [hunk 2]
```diff
@@ -33,6 +33,9 @@ internal class AudioSessionManager(
             FeatureFlags.BluetoothAudio.active &&
                 (bluetoothAudioProxy?.connectedDevices?.size ?: 0 > 0)
 
+    private val bluetoothDeviceName
+        get() = bluetoothAudioProxy?.connectedDevices?.firstOrNull()?.name ?: ""
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt:37]: Should we use context.getString(R.string.azure_communication_ui_setup_audio_device_bluetooth) here instead of empty string. This will allow us not to checks in the views for empty string.

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarViewModel.kt [hunk 3]
```diff
@@ -26,6 +26,9 @@ internal class SetupControlBarViewModel(
     private lateinit var audioDeviceSelectionStatusStateFlow: MutableStateFlow<AudioDeviceSelectionStatus>
     private lateinit var callingStatusStateFlow: MutableStateFlow<CallingStatus>
 
+    // Does not need StateFlow as tied to AudioDeviceSelectionStatus updates
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarViewModel.kt:29]: I think we can remove this comment

**Comment by iaulakh** [comment]
We can write unit tests as most of logic is testable

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/calling/CallingViewModel.kt (hunk 1 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/calling/CallingViewModel.kt]: A dedicated Bluetooth class has been added, and the ViewModels for SetupControlBarView and AudioDeviceListView have been updated to bring in this parent class (instead of each of its fields independently).

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt (hunk 2 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt]: Dropped getString in the view, push the device name to Redux from AudioDeviceManager now, which is pulled into the VM.

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarViewModel.kt (hunk 3 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarViewModel.kt:28]: Deleted this member field (and comment), as it all comes through AudioState and BluetoothState now.

**Comment by ahammer** [comment]
I'll update with tests

**Review by iaulakh — CHANGES_REQUESTED** [review]
Hi Adam, while testing PR I noticed that, when I tried to launch composite from sample app, immediately I received a permission dialog for "record audio" and application went to background.
I do not notice this behaviour before when audio permissions was requested. 
Screen Shot 2022-03-06 at 7 28 30 AM

**Comment by ahammer** [comment]
> Hi Adam, while testing PR I noticed that, when I tried to launch composite from sample app, immediately I received a permission dialog for "record audio" and application went to background. I do not notice this behaviour before when audio permissions was requested. Screen Shot 2022-03-06 at 7 28 30 AM

bluetooth_connect is a runtime permission in API 31 and above, I've had to boost target SDK to 31 to access the bluetooth_connect feature, and added it to the permissions that get requested for audio.

File: azure-communication-ui/azure-communication-ui/build.gradle [hunk 4]
```diff
@@ -9,13 +9,13 @@ apply from: file('publish.gradle')
 apply from: file('jacoco.gradle')
 
 android {
-    compileSdkVersion 30
+    compileSdkVersion 31
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/build.gradle:12]: Adam, we can do this in separate PR. We need to update build tools to version 31 for this change.

**Review by iaulakh — CHANGES_REQUESTED** [review]
Adam, noticed these issues bug:
1 --------------------------
Uninstall app
Connect bluetooth headset
Install app
On setup screen. no bluetooth device is displayed
2 -------------------------
Only Speaker and bluetooth headset is displayed as option, no Android
3 -------------------------
Change audio option to Android
Connect bluetooth device
Turn off bluetooth device
Audio option is changed to speaker (Expected: The previous device should be selected)

File: azure-communication-ui/azure-communication-ui/build.gradle (hunk 4 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/build.gradle:12]: It's required for this PR, so it'll have to be merged first.

**Comment by ahammer** [comment]
@iaulakh 31 was merged in a separate commit and merged into this, and tests added for the device fallback action, I think it should be good again for review. Tested in standup this morning some more as well.

**Review by iaulakh — DISMISSED** [review]
@ahammer noticed that, if bluetooth device is on, on switching to speaker, turning on bluetooth device again, audio switched to bluetooth. I think, we need to switch only when Android is selected, not speaker. Please confirm.

Please also follow up with added redux state with iOS team. We can create refactoring task in future if required.

**Comment by ahammer** [comment]
I talked to @iaulakh a bit more, and upon the realization that adding actions will span more alignment and team discussions, I'm going to amend this tomorrow to drop the new actions and state (I realize now I can handle in Manager without redux state changes). This means less sync discussions and overhead, and lower code and redux complexity.

I will keep the bluetooth info in state (e.g. device name), but drop the last device state and disconnect bluetooth action and re-use the existing action and move state to the manager.

**Comment by ahammer** [comment]
This PR is clean again now, the additional actions have been dropped, with the exceptions of bluetooth availability and device name state which are required.

**Review by pavelprystinka — DISMISSED** [review]

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt [hunk 5]
```diff
@@ -35,7 +36,16 @@ internal class AudioSessionManager(
...
+                ?: context.getString(R.string.azure_communication_ui_setup_audio_device_bluetooth)
+        } catch (exception: SecurityException) {
+            context.getString(R.string.azure_communication_ui_setup_audio_device_bluetooth)
+        }
+
     private var previousAudioDeviceSelectionStatus: AudioDeviceSelectionStatus? = null
+    private var priorToBluetoothAudioSelectionStatus: AudioDeviceSelectionStatus? = null
 
     suspend fun start() {
         BluetoothAdapter.getDefaultAdapter()?.run {
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt:51]: getDefaultAdapter() is deprecated in API31
https://developer.android.com/reference/android/bluetooth/BluetoothAdapter#getDefaultAdapter()

**Review by phileo99 — CHANGES_REQUESTED** [review]

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt (hunk 5 above)
Comment (phileo99) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt:51]: acutally, @ahammer can you create a separate refactor ticket for the deprecated function, and I will approve this one

Comment (phileo99) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt:51]: acutally, @ahammer can you create a separate refactor ticket for the deprecated function, and I will approve this one

**Review by phileo99 — APPROVED** [review]
manually tested, passed Espresso test suite.
Deprecated function call to be dealt with in separate ticket
👍

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt (hunk 5 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/manager/AudioSessionManager.kt:51]: Hmmm OK, I'll take a look.

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarView.kt [hunk 6]
```diff
@@ -129,6 +129,10 @@ internal class SetupControlBarView : LinearLayout {
         audioDeviceButton.refreshDrawableState()
     }
 
+// <<<<<<< HEAD
```
Comment (iaulakh) [inline azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarView.kt:132]: We can remove this comment Adam

**Review by iaulakh — APPROVED** [review]

File: azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarView.kt (hunk 6 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/main/java/com/azure/android/communication/ui/presentation/fragment/setup/components/SetupControlBarView.kt:132]: I'll follow up in the second PR, (with wired headphone) don't want to clear approvals and chase for a day
